src/actions/pkg.cr

```diff
@@ -1,45 +1,137 @@
+        error = NoPkgsToClean.new context
+        error.cause = bug
+        raise error
+      end
+      raise NoPkgsToClean.new context
     end
-    raise Unauthorized.new context
-  end
-  # Query information about a given package
-  relative_get "/:id/query" do |context|
-    if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: Query information about the given package
-      next context
+
+    # List built packages
+    relative_get nil do |context|
+      if context.current_user? && Config.has_access? context, Access::Read
+        # Build JSON directly to the HTTP response IO
+        build_json context.response do |json|
+          json.array do
+            # build an array of all the package names
+            get_prefix_or_default(from: context).each_pkg { |pkg| json.string pkg.package }
+          end
+        end
+        next context
+      end
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Delete a given package
-  relative_delete "/:id/delete" do |context|
-    if context.current_user? && Config.has_access? context, Access::Delete
-      # TODO: Query information about the given package
-      next context
+    # Clean unused built packages
+    relative_delete "/clean" do |context|
+      if context.current_user? && Config.has_access? context, Access::Delete
+        clean_unused_packages(context).to_json context.response
+        context.response.flush
+        next context
+      end
+      raise Unauthorized.new context
     end
-    raise Unauthorized.new context
-  end
-  # Build a package, returning the ID of the built image, and perhaps a status
-  # message? We could also use server-side events or a websocket to provide the
-  # status of this action as it occurs over the API, rather than just returning
-  # a result on completion.
-  relative_post "/:id/build" do |context|
-    if context.current_user? && Config.has_access? context, Access::Create
-      # TODO: build the package based on the submitted configuration
+    # Query information about a given package
+    relative_get "/:id/query" do |context|
+      if context.current_user? && Config.has_access? context, Access::Read
+        package_name = URI.unescape context.params.url["id"]
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        tag = context.params.query["tag"]?.try { |t| URI.unescape t }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = DPPM::Prefix::Pkg.create get_prefix_or_default(from: context),
+          package_name,
+          version,
+          tag
+        raise NoSuchPackage.new context, package_name if selected_pkg.nil?
+        # Stop here ^^ unless the package whose name was the :id URL parameter
+        # was found and we can query it
+        if key = context.params.query["get"]?
+          data = selected_pkg.get_config? key
+          if data.nil?
```

Comment:
Use `ConfigKeyError`.
A config key can be present, with a nil value.

Reply:
oof good catch, thank you